cloud-formation/python/geolactor-lambda.py

```diff
@@ -0,0 +1,124 @@
+"""
+Function: Geolocator-lambda
+Tasks: 
+* Handle the url request from the API to identify,
+  validate and assert the input query and additional parameters
+
+* Retrieve the set of attributes and methods associated 
+  with each service required in the url request
+* Build and execute a new REST request for each associated service
+* Retrieve and return the normalized payload form the call
+"""
+import json
+import urllib.request
+
+class URL_object:
+    def __init__(self, name, url_base, address, dict_params={}):
+        self.name = name
+        self.url = url_base
+        self.params = dict_params
+        self.address = address
+    
+    """
+    def getUrl(self, query_str):
+        qry = self.url + "?"+self.address+"="+query_str
+        for key in self.params:
+            value = self.params[key]
+            param_str="&"+key+"="+value
+            qry += param_str
+        return qry
+    """
+    
+    def get_url(self):
+        return self.url
+                
+    def get_q(self):
+        return self.address
+
+def read_services():
+    # TODO: save Objects parameters and schema in database
+    services = {}
+    services["geonames"] = URL_object("geonames","https://geogratis.gc.ca/services/geoname/_PARAM1_/geonames.json","q")
+    services["nominatim"] = URL_object("nominatim","https://nominatim.openstreetmap.org/search", "q",{"format":"jsonv2"})
+    #services["google"] = URL_object("google","https://maps.googleapis.com/maps/api/geocode/json", "address", {"key":"AIzaSyASQcYTDCw4fRr_GY5WHxIAqeTsDmvAh_8"})
+    return services
+
+
+def catch_unknown_param(entered_params, valid_params):
+    unknown_params = ','.join(list(set(entered_params) - set(valid_params)))
+    if unknown_params:
+        error_message = "paramaters not required: " + unknown_params
+        raise Exception(error_message)
+
+def validate_query_string(queryString):
+    params_list = {}
+    # List of valids must be placed and retrieved from elsewhere
+    valid_parms = ['q','keys','lang']
+    valid_langs = ['en','fr']
+    valid_services = ['geonames','nominatim']
```

Comment:
Key list is built from the neame of metedata file in the S3 bucket. Not hard coded